### PR TITLE
Sidesheet

### PR DESCRIPTION
**Pull request by Gustav-Eikaas**
# Description

A sidesheet controller heavily inspired from experience gained throughout the Castberg Project portal. The controller supports closing the sidesheet without removing the item. I felt like this was a better approach than the minimize functionality we had in Castberg Portal. Everything is tested and seems to work like it should. There is a function for removing the item from the controller aswell as a function for closing the sidesheet. I want them to be two different functions, but seeing how tedious it is to close and remove the item. I made a function for that aswell.

Completes: AB#FILL_IN_YOUR_ISSUE_ID

## Checklist

- [x] I have performed a self-review of my own code.
- [ ] I have linked my DevOps task using the AB# tag.
- [x] My code is easy to read.
- [x] Documentation and comments is added where needed.
- [x] My code is covered by tests.

### CODE REVIEW

File: packages/sidesheet/src/lib/test/onItemChanges.spec.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+const assertName = 'test';
+
+const mockFunction = jest.fn();
+
+describe('Item in sidesheet changes', () => {
+    const controller = new SidesheetController();
+    it('Should fire callback when item changes', () => {
+        controller.onItemChanges((s) => {
+            expect(assertName).toBe(assertName);
```
Comment (Gustav-Eikaas) [inline packages/sidesheet/src/lib/test/onItemChanges.spec.ts:11]: expect(s).toBe(assertName)

File: packages/sidesheet/src/lib/classes/sidesheetController.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    /** Function that both closes the sidesheet and removes the item */
+    closeAndRemoveItem = () => {
+        this.setSidesheetState('Closed');
+        this.setItem(undefined);
+    };
+
+    /**
+     * Register a callback to be called upon when the sidesheet state changes
+     * @param cb
```
Comment (Gustav-Eikaas) [inline packages/sidesheet/src/lib/classes/sidesheetController.ts:38]: tsdoc

File: packages/sidesheet/src/lib/classes/sidesheetController.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+            unsubscribe: () => {
+                this.onSidesheetStateChangedCallbacks =
+                    this.onSidesheetStateChangedCallbacks.filter((s) => s.id !== id);
+            },
+        };
+    };
+
+    /**
+     * Function for opening or closing the sidesheet
+     * @param state
```
Comment (Gustav-Eikaas) [inline packages/sidesheet/src/lib/classes/sidesheetController.ts:55]: tsdoc

File: packages/sidesheet/src/lib/classes/sidesheetController.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        this.sidesheetState = state;
+
+        if (!preventCallbacks) {
+            this.onSidesheetStateChangedCallbacks.forEach(({ callback }) => callback(state, this));
+        }
+    };
+
+        /**
+         * Function for setting a new item for the sidesheet controller to hold
+         * @param item
```
Comment (Gustav-Eikaas) [inline packages/sidesheet/src/lib/classes/sidesheetController.ts:70]: tsdoc

File: packages/sidesheet/src/lib/classes/sidesheetController.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    setItem = (item: TData | undefined, preventCallbacks?: boolean) => {
+        this.item = item;
+        if (!preventCallbacks) {
+            this.onItemChangedCallbacks.forEach(({ callback }) => callback(item, this));
+        }
+    };
+
+
+    /**
+     * Callback to be called on whenever the item changes
```
Comment (Gustav-Eikaas) [inline packages/sidesheet/src/lib/classes/sidesheetController.ts:82]: tsdoc

File: packages/sidesheet/src/lib/hooks/useSidesheetState.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import { useEffect, useState } from "react";
+import { SidesheetController } from "../classes";
+
+export function useSidesheetState<TItem>({onSidesheetStateChanged, isSidesheetOpen}: SidesheetController<TItem>){
+    const [isOpen, setIsOpen] = useState<boolean>(isSidesheetOpen());
+
+
+    useEffect(() =>{
+        const { unSubscribe } = onSidesheetStateChanged(() => {
```
Comment (Gustav-Eikaas) [inline packages/sidesheet/src/lib/hooks/useSidesheetState.ts:9]: unsubscribe

**Review by jaysencpp — APPROVED** [review]

**Pull request merged**